### ISSUE TITLE
Floating windows spawn on screen 1

### ISSUE THREAD

**Issue by tcld** [qtile/qtile#1913]
# Issue description

I have two screens attached to my machine. When opening certain floating windows, they appear on the wrong screen. Specifically imagine the following situation:

Group A is active on screen 1, group B is active on screen 2. Group B/screen 2 has focus. I launch a program and it shows up on screen 1, but it is managed by group B. If there are any windows in group A, it is then not possible to move the floating window around since the windows in group A grab the focus instead.

From what I have seen, this only applies to windows defined as floating through the `floating_layout` variable (windows put into floating mode manually don't show this problem). After digging in the code a bit it looked as if the screen offset wasn't added to the `float_x` and `float_y`-variables (like it is done in window.place()), but I didn't see an easy way to fix or even confirm that.

# Qtile version

https://github.com/qtile/qtile/commit/f7c212db2abce06f78c23f3aaa4e2fa03affcdb1

# Configuration

Just my floating configuration here. If anybody thinks that another section might be helpful, I could add it.

```python
    floating_layout = layout.Floating(
        float_rules=[
            # Run the utility of `xprop` to see the wm class and name of an X client.
            {'wmclass': 'confirm'},
            {'wmclass': 'dialog'},
            {'wmclass': 'download'},
            {'wmclass': 'error'},
            {'wmclass': 'file_progress'},
            {'wmclass': 'notification'},
            {'wmclass': 'splash'},
            {'wmclass': 'toolbar'},
            {'wmclass': 'confirmreset'},  # gitk
            {'wmclass': 'makebranch'},  # gitk
            {'wmclass': 'maketag'},  # gitk
            {'wname': 'branchdialog'},  # gitk
            {'wname': 'pinentry'},  # GPG key password entry
            {'wmclass': 'ssh-askpass'},  # ssh-askpass
            {'wmclass': 'lxpolkit'}, # credential management by authentication agent
            {'wmclass': 'Xephyr'},  #window manager testing w/ Xephyr
            {'wname': 'Compress'}, # compression extension for file browsers
            {'wmclass': 'veracrypt'},
        ],
        no_reposition_match=Match(wm_class=['Gsimplecal']),  # do not reposition windows to center of screen (leave it to them instead)
        **layout_theme
    )
```

**Comment by tcld** [comment]
I might have found the cause for this:
In PR #1710 [a check](https://github.com/qtile/qtile/commit/0f886606e1f35226b325f5405d43946976eae831#diff-475097a9172bb7cb3cc2bc9fe378a383acf1231b6a6bc1ec048e764af439eddcR221) was added in the floating layouts `configure`-method to see whether a window had been placed by the user. The actual problem might be caused by a later commit, but with the current master I was not able to find even a _single_ window that didn't pass that check. Therefore the floating layout right now doesn't lay out any windows at all!

@tych0, @ramnes : Removing that one check seems to fix all the issues I have observed with floating windows (weirdly enough even #1893, it seems). But maybe the corresponding flags are wrongly set, then a fix would have to be applied somewhere else (I am totally in the dark as to where, though).

**Comment by tcld** [comment]
Without the check, moving a manually floated window across screen borders does move the window into another group. Before, this would cause some weird issues similar to what I described above.

**Comment by tcld** [comment]
I did a few more tests. It seems that, upon first reaching the method `core.manager.map_window` (i.e. before even the `_Window`-constructor is called), the `WM_NORMAL_HINTS` for the Nextcloud main window are already at value 515, which includes `USPosition` (1). The flags for a window defined in a floating rule (in this case Veracrypt) start out at 820, which includes `PPosition` (4), and the flags for a floating child window of VirtualBox even start out with both. So any window that starts floating has at least one of the flags already set and can thus not be managed by the floating layout.

Maybe somebody has some advice where to continue looking for the source of this. :)

**Comment by tcld** [comment]
I investigated a little further, in case it is even helpful anymore: I added the following block of code in a few places:

```python
try:
    wm_normal_hints = window.get_property(  # `window` is an xcbq.Window here
        "WM_NORMAL_HINTS",
        xcffib.xproto.GetPropertyType.Any
    )
    if wm_normal_hints:
        atom_list = wm_normal_hints.value.to_atoms()
        logger.warning('%s: USPosition - %s, PPosition - %s', window.get_name(),
            atom_list[0] & xcbq.NormalHintsFlags['USPosition'] != 0,
            atom_list[0] & xcbq.NormalHintsFlags['PPosition'] != 0
        )
except Exception as e:
    logger.error(e)
```

Even checking the flags as early as in `backend.x11.xcore.handle_MapRequest`, for many programs* the flags are already set. To me it looks like we can't rely on those two flags to tell us whether a window was _actually_ positioned by the user or Qtile.
For all of these windows, if they were to be added to a floating layout, they could cause problems like described here (#1913) or #1893 (this affects child-windows *and* popups). I have no idea how to work around this and I'll stop looking further into this for now.

*= LibreOffice, qutebrowser, Veracrypt, Nextcloud (the only ones I found so far where neither flag is set are Firefox and Alacritty).

**Comment by m-col** [comment]
This got me wondering... do we ever need to set the position of a floating
window the moment it starts floating? I've found the current behaviour fine,
which if, as your saying, has always just left the window do whatever (including
nothing), might be quite good.

For instance, making a tiled window floating doesn't change the geometry at all,
unless the window has these user positioning flags, and that might be... good?

**Comment by tcld** [comment]
Well, the window should at least be moved to the current screen. And at that point we are touching its coordinates anyway.

**Comment by m-col** [comment]
Oh yeah, got ahead of myself. We could just always get the coordinates, add the screen's coordinates, then reposition with those?

**Comment by tcld** [comment]
That's about what is being done in `window._Window.place`. So I guess ideally the floating layout just stops touching those values (it'll still have to take care of z, of course) and we could hope that works well for most situations.

**Comment by tcld** [comment]
I just gave this a try by commenting out [a few](https://github.com/qtile/qtile/blob/master/libqtile/layout/floating.py#L207) [lines](https://github.com/qtile/qtile/blob/master/libqtile/layout/floating.py#L234) in `floating.py`. Surprisingly enough, I only found one floating window that didn't appear exactly where I would have expected it; that window appeared at (0, 0) instead. Even child-windows (I only tested VirtualBox) appeared centered on their parent. So it seems that leaving the windows alone is a decent approach. the (0, 0)-case could (and probably should) be caught.

**Comment by m-col** [comment]
> Also, generally it might be nice to respect USPosition and PPosition instead of ignoring them entirely. Maybe we could even leverage those for use by Qtile somehow.

Sorry, I might not have been clear. Looks like you discovered it in that second comment though: Qtile doesn't necessarily have to deal with the user position hints at all, because the X server will still place windows where they want. TBH I'm surprised that the screen would need accounting for in Qtile... surely USPosition information used directly by the X server would account for screens?

If it makes things more clean, we could make all of the arguments to _WIndow.place optional kwargs so we only touch the attributes we need to.

**Comment by tcld** [comment]
I tried to adress this in #1943, but I didn't touch `_Window.place` while doing so. Maybe you can have a look and tell me what further changes you would have had in mind.

**Comment by tcld** [comment]
Closed via #1943.
